lib/internal/trace_events_async_hooks.js

```diff
@@ -64,4 +64,4 @@ function createHook() {
   });
 }
 
-module.exports = { createHook };
+module.exports = createHook;
```

Comment:
At least make this a singleton. Meaning `module.exports = createHook();`

Reply:
I had it that way in the original version but I've got concerns over the internal state tracking... Specifically, we end up with a memory leak if tracing is dynamically disabled and we still have items in the map that are waiting for a destroy. Doing it this way resets the internal state. Will think of another way of doing that.

Reply:
One approach we could take is to add a reset function to the async_hook instance. I could do that in a one off way or we could make it part of the async_hook API. Internally, the async_hook can call reset when disable is called.

Reply:
@jasnell I think it is best to do it in a "one-off way". I would even consider just exposing a `enable` and a `disable` function, that simply relays to `async_hooks` and clears the state.

In general, I think `async_hooks` should be as state-free as possible. As it can get very heavy performance wise and is easily implemented in userland.